v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/SpannerToSourceDb.java

```diff
@@ -299,6 +321,38 @@ public interface Options extends PipelineOptions, StreamingOptions {
     String getDeadLetterQueueDirectory();
 
     void setDeadLetterQueueDirectory(String value);
+
+    @TemplateParameter.Integer(
+        order = 21,
+        optional = true,
+        description = "Dead letter queue maximum retry count",
+        helpText =
+            "The max number of times temporary errors can be retried through DLQ. Defaults to 500.")
+    @Default.Integer(500)
+    Integer getDlqMaxRetryCount();
+
+    void setDlqMaxRetryCount(Integer value);
+
+    @TemplateParameter.Enum(
+        order = 22,
+        optional = true,
+        description = "Run mode - currently supported are : regular or retryDLQ",
+        enumOptions = {@TemplateEnumOption("regular"), @TemplateEnumOption("retryDLQ")},
+        helpText = "This is the run mode type, whether regular or with retryDLQ.")
```

Comment:
The "retryDLQ" is only for retrying the severe errors. This is same as the forward migration template. In the retryDLQ mode, the source of input is only the DLQ records and not change streams.

Reply:
I think we discussed this earlier but just bringing it up again in case we want to revisit. 

Severe DLQ will contain a finite set of entries - is a streaming job a good fit for retrying DLQ entries? The customer will not have an easy way to determine if all the retries have been processed since the job never finishes..

Maybe it makes sense to have a DLQ batch job which can retry bulk, forward migration and reverse replication DLQ messages based on the parameters - which should be the "retryType" and the "GCS location"  (rest would be the usual ones)

WDYT? 

CC: @bharadwaj-aditya

We are trying to retry bulk DLQ via the retryDLQ mode in streaming job which is also counter intuitive IMO as a long-term solution.

Reply:
I also think it overloads the template implementation since it now effectively supports two sources - Spanner Changestreams and GCS.

Reply:
In a purist sense, yes the records are finite. 

But the benefits of re-using an existing template logic are substantial:  which is - a) only the source of input changes and rest of the logic remains same - easier to reason and maintain b) No need to implement,maintain and test multiple templates c) easier for the user also - since only change is the run mode between the two templates.
User would anyway monitor the metrics to check if all records got reprocessed - so stopping the template should not be a big burden.


I am not in favor for a monolith template that will invoke bulk, reverse and forward logic. From separation of concerns pov.  separate templates makes sense - since each has its own distinct processing logic - be it regular or retry.